File: js/recursion/subsetSum.js
/*
    Problem Statement:
        - given an array and a target sum, print out every
        possible subset of the input array that adds up to the
        target sum

    approach:
        - generate subsets recursively,
            - reduce target sum by value of next element in array
            - if target sum == 0, we have found a valid subset
                - print it out and return (break recursive call)
            - else if target sum < 0, or index exceeds bounds of our array, 
            we have added an element to our subset that now makes it invalid
                - return 
            - else we have two choices, add next element to subset or not,
                - recursive call where we chose to add it to subset in one call
                - and another recursive call where we do not add it to the subset

*/

function helper(array, subset, targetSum, index) {
    // perfect sum found when targetSum = 0
    // print and return
    if (targetSum === 0) {
        console.log(subset)
        return
    }
    
    // if we have exceeded our target sum (< 0) or 
    // could exceed the array length (index === array.length)
    // we no longer need to go down this path, return
    if (targetSum < 0 || index === array.length) {
        return
    }   

    // recursive case where we do add element to subset
    subset.push(array[index])
    helper(array, subset, targetSum - array[index], index + 1);

    // recursive case where we do NOT add element to subset
    subset = subset.filter(element => element !== array[index]) // only keep elements not equal to the one at given index
    helper(array, subset, targetSum, index + 1)


}

function subsetSum(array, targetSum) {
    let subset = []
    return helper(array, subset, targetSum, 0)
}


let a = [1, 2]
let target = 3
// subsetSum(a, target)

File: python/recursion/subsetSum.py
# subset-sum problem

# given s = {1, 2, 3} and target = 3,
# generate all the subsets whose sum == target

# time-complexity: O(2^n)
#   -> just like combinations; every node i doing a constant amount of work
#       then the complexity would be proportional to the 
#       total amount of nodes in the tree ~ O(2^n)
#  
# space-complexity: O(n) 
#   -> proportional to maximum call stack, O(n)

def helper(givenArray, subset, runningSum, index):
    if runningSum == 0: # we have found a subset that adds to target sum
        print(subset)
        return
    if runningSum < 0 or index == len(givenArray): 
        return
    # add element at given index to subset
    subset.append(givenArray[index])
    # recursive call with included element
    helper(givenArray, subset, runningSum - givenArray[index], index + 1)
    # dont add element at given index to subset
    subset.remove(givenArray[index])
    # recursive call without current element
    helper(givenArray, subset, runningSum, index + 1)


def subsetSum(givenArray, targetSum):
    subset = []
    return helper(givenArray, subset, targetSum, 0)
    
def test():
    a = [1, 2, 4, 20, 5, 19, 6, 14, 13, 21]
    target = 25

    subsetSum(a, target)

test()